File: packages/snow_draw_core/lib/draw/types/draw_point.dart
import 'dart:math';

import 'package:meta/meta.dart';

import 'draw_rect.dart';

@immutable
class DrawPoint {
  const DrawPoint({
    required this.x,
    required this.y,
    this.pressure = 0.0,
    this.timestamp = 0,
  });

  factory DrawPoint.fromPoint(Point<double> point) =>
      DrawPoint(x: point.x, y: point.y);
  final double x;
  final double y;

  /// Stylus / pointer pressure in the range 0..1.
  ///
  /// A value of 0 means pressure is unknown or unavailable.
  final double pressure;

  /// Monotonic timestamp in microseconds, used for velocity calculation.
  ///
  /// A value of 0 means the timestamp is unavailable.
  final int timestamp;

  static const zero = DrawPoint(x: 0, y: 0);

  DrawPoint copyWith({
    double? x,
    double? y,
    double? pressure,
    int? timestamp,
  }) => DrawPoint(
    x: x ?? this.x,
    y: y ?? this.y,
    pressure: pressure ?? this.pressure,
    timestamp: timestamp ?? this.timestamp,
  );

  DrawPoint translate(DrawPoint position) =>
      DrawPoint(x: x + position.x, y: y + position.y);

  DrawPoint operator +(DrawPoint other) =>
      DrawPoint(x: x + other.x, y: y + other.y);

  DrawPoint operator -(DrawPoint other) =>
      DrawPoint(x: x - other.x, y: y - other.y);

  DrawPoint operator -() => DrawPoint(x: -x, y: -y);

  /// Whether this point carries valid pressure data.
  bool get hasPressure => pressure > 0;

  Point<double> toPoint() => Point(x, y);

  DrawRect toRect(DrawPoint other) {
    var minX = x;
    var maxX = other.x;
    var minY = y;
    var maxY = other.y;

    if (minX > maxX) {
      final temp = minX;
      minX = maxX;
      maxX = temp;
    }

    if (minY > maxY) {
      final temp = minY;
      minY = maxY;
      maxY = temp;
    }

    return DrawRect(minX: minX, minY: minY, maxX: maxX, maxY: maxY);
  }

  /// Computes the Euclidean distance to [other].
  double distance(DrawPoint other) => sqrt(distanceSquared(other));

  /// Computes the squared distance to [other], useful for comparisons without
  /// calling `sqrt`.
  double distanceSquared(DrawPoint other) {
    final dx = x - other.x;
    final dy = y - other.y;
    return dx * dx + dy * dy;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      (other is DrawPoint &&
          other.x == x &&
          other.y == y &&
          other.pressure == pressure);

  @override
  int get hashCode => Object.hash(x, y, pressure);

  @override
  String toString() =>
      'DrawPoint(x: $x, y: $y${hasPressure ? ', p: $pressure' : ''})';
}

File: packages/snow_draw_rust_engine/src/draw/types/draw_point.rs
#![allow(dead_code)]

use crate::draw::types::draw_rect::DrawRect;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::hash::{Hash, Hasher};
use std::ops::{Add, Div, Mul, Neg, Sub};

/// Immutable 2D point used by draw geometry.
#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize)]
pub struct DrawPoint {
    pub x: f64,
    pub y: f64,

    /// Stylus / pointer pressure in the range `0..=1`.
    pub pressure: f64,

    /// Monotonic timestamp in microseconds.
    pub timestamp: i64,
}

impl DrawPoint {
    pub const ZERO: Self = Self {
        x: 0.0,
        y: 0.0,
        pressure: 0.0,
        timestamp: 0,
    };

    pub const fn new(x: f64, y: f64) -> Self {
        Self {
            x,
            y,
            pressure: 0.0,
            timestamp: 0,
        }
    }

    pub const fn with_pressure_and_timestamp(
        x: f64,
        y: f64,
        pressure: f64,
        timestamp: i64,
    ) -> Self {
        Self {
            x,
            y,
            pressure,
            timestamp,
        }
    }

    pub fn copy_with(
        self,
        x: Option<f64>,
        y: Option<f64>,
        pressure: Option<f64>,
        timestamp: Option<i64>,
    ) -> Self {
        Self {
            x: x.unwrap_or(self.x),
            y: y.unwrap_or(self.y),
            pressure: pressure.unwrap_or(self.pressure),
            timestamp: timestamp.unwrap_or(self.timestamp),
        }
    }

    pub fn translate(self, position: DrawPoint) -> Self {
        self + position
    }

    pub fn has_pressure(self) -> bool {
        self.pressure > 0.0
    }

    pub fn to_tuple(self) -> (f64, f64) {
        (self.x, self.y)
    }

    pub fn to_rect(self, other: DrawPoint) -> DrawRect {
        DrawRect::new(
            self.x.min(other.x),
            self.y.min(other.y),
            self.x.max(other.x),
            self.y.max(other.y),
        )
    }

    /// Computes the Euclidean distance to `other`.
    pub fn distance(self, other: DrawPoint) -> f64 {
        self.distance_squared(other).sqrt()
    }

    /// Computes squared distance to `other` without applying `sqrt`.
    pub fn distance_squared(self, other: DrawPoint) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        dx * dx + dy * dy
    }
}

impl Add for DrawPoint {
    type Output = DrawPoint;

    fn add(self, rhs: Self) -> Self::Output {
        DrawPoint::new(self.x + rhs.x, self.y + rhs.y)
    }
}

impl Sub for DrawPoint {
    type Output = DrawPoint;

    fn sub(self, rhs: Self) -> Self::Output {
        DrawPoint::new(self.x - rhs.x, self.y - rhs.y)
    }
}

impl Mul<f64> for DrawPoint {
    type Output = DrawPoint;

    fn mul(self, rhs: f64) -> Self::Output {
        DrawPoint::new(self.x * rhs, self.y * rhs)
    }
}

impl Div<f64> for DrawPoint {
    type Output = DrawPoint;

    fn div(self, rhs: f64) -> Self::Output {
        DrawPoint::new(self.x / rhs, self.y / rhs)
    }
}

impl Neg for DrawPoint {
    type Output = DrawPoint;

    fn neg(self) -> Self::Output {
        DrawPoint::new(-self.x, -self.y)
    }
}

impl From<(f64, f64)> for DrawPoint {
    fn from(value: (f64, f64)) -> Self {
        DrawPoint::new(value.0, value.1)
    }
}

impl PartialEq for DrawPoint {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.y == other.y && self.pressure == other.pressure
    }
}

impl Eq for DrawPoint {}

impl Hash for DrawPoint {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.x.to_bits().hash(state);
        self.y.to_bits().hash(state);
        self.pressure.to_bits().hash(state);
    }
}

impl fmt::Display for DrawPoint {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.has_pressure() {
            write!(
                f,
                "DrawPoint(x: {}, y: {}, p: {})",
                self.x, self.y, self.pressure
            )
        } else {
            write!(f, "DrawPoint(x: {}, y: {})", self.x, self.y)
        }
    }
}
